### ISSUE TITLE
question about fine tuning falcon

### ISSUE THREAD

**Issue by charlesmartin14** [georgian-io/LLM-Finetuning-Hub#57]
Hello
  I ran the falcon classification task uaing the following command:
`!python falcon_classification.py --lora_r 64 --epochs 1 --dropout 0.1 # finetune Falcon-7B on newsgroup classification dataset
`
  Upon inspecting the model, I find that many of the layers  are full rank and not the lower rank

```
from transformers import AutoModelForCausalLM
model = AutoModelForCausalLM.from_pretrained("experiments/classification-sampleFraction-0.99_epochs-1_rank-64_dropout-0.1/assets")
 
```
Here is a screenshot showing this

<img width="1255" alt="Screenshot 2023-10-24 at 8 41 42 PM" src="https://github.com/georgian-io/LLM-Finetuning-Hub/assets/498448/fb224b07-47fb-449c-9cdb-e5645eccbe62">

Is this expected behvior ?

**Comment by RohitSaha** [comment]
Great question @charlesmartin14 .

When you do `model = AutoModelForCausalLM.from_pretrained("experiments/classification-sampleFraction-0.99_epochs-1_rank-64_dropout-0.1/assets")`, the base model, i.e., Falcon-7B gets loaded, **_without_** the LoRA config. 

So yes, what you see in the screenshot is expected behaviour.

In order to load the low rank matrices, you will have to also do: 

`model = PeftModel.from_pretrained(model, peft_model_id)`, 

where peft_model_id= "experiments/classification-sampleFraction-0.99_epochs-1_rank-64_dropout-0.1/assets"

The above line will load the low rank matrices trained via LoRA.

Following is the full sequence of steps:

```
    config = PeftConfig.from_pretrained(peft_model_id)

    bnb_config = BitsAndBytesConfig(
        load_in_4bit=True,
        bnb_4bit_quant_type="nf4",
        bnb_4bit_compute_dtype=torch.float16,
    )

    model = AutoModelForCausalLM.from_pretrained(
        config.base_model_name_or_path,
        quantization_config=bnb_config,
        trust_remote_code=True,
    )
    model = PeftModel.from_pretrained(model, peft_model_id)
    model.eval()
```

**Comment by charlesmartin14** [comment]
Let me clarify just tp be sure

"The above line will load the low rank matrices trained via LoRA."

I am looking to load *only* the LoRA delta/update and not the full model + LoRA

**Comment by benjaminye** [comment]
Hi @charlesmartin14,


It seems like `AutoModelForCausalLM.from_pretrained` does load in LoRA weights by default. 

<img width="796" alt="image" src="https://github.com/georgian-io/LLM-Finetuning-Hub/assets/17937357/4018137b-4df7-4a94-a823-a639707f634a">

As seen from above, the loaded model has both the Base Model layers, as well as LoRA layers.


To inspect LoRA weights, you can use the nifty utility function `get_peft_model_state_dict`.

Here's an example following your code:

```python
from peft import get_peft_model_state_dict

for layer_id, layer in get_peft_model_state_dict(model).items():
    print(layer_id, layer.shape)
```

```
>>>
transformer.h.0.self_attention.query_key_value.lora_A.weight torch.Size([64, 4544])
transformer.h.0.self_attention.query_key_value.lora_B.weight torch.Size([4672, 64])
transformer.h.0.self_attention.dense.lora_A.weight torch.Size([64, 4544])
transformer.h.0.self_attention.dense.lora_B.weight torch.Size([4544, 64])
...
```

As you can see from the output, each LoRA matrices `B` and `A` are of correct shapes for a rank-64 decomposition.

**Comment by benjaminye** [comment]
Closed due to inactivity.
